File: backend/system/operation_recorder.py
# -*- coding: utf-8 -*-
"""
操作捕捉API
受 GPL v3.0 保护

录制/回放/导出。
"""

from flask import Blueprint, request, jsonify, send_file

operation_recorder_bp = Blueprint('operation_recorder', __name__)

_demo_runner = None
_export_engine = None


def set_deps(demo_runner, export_engine=None):
    """注入依赖"""
    global _demo_runner, _export_engine
    _demo_runner = demo_runner
    _export_engine = export_engine


@operation_recorder_bp.route('/api/recorder/start', methods=['POST'])
def start_record():
    """开始录制"""
    if _demo_runner:
        result = _demo_runner.record_action({'type': 'start'})
        return jsonify({'success': True, 'data': result})
    return jsonify({'success': False, 'message': '演示执行器未初始化'}), 500


@operation_recorder_bp.route('/api/recorder/stop', methods=['POST'])
def stop_record():
    """停止录制"""
    if _demo_runner:
        result = _demo_runner.stop_record()
        return jsonify({'success': True, 'data': result})
    return jsonify({'success': False}), 500


@operation_recorder_bp.route('/api/recorder/records', methods=['GET'])
def list_records():
    """录制列表"""
    if _demo_runner:
        return jsonify({'success': True, 'data': _demo_runner.get_records()})
    return jsonify({'success': True, 'data': []})


@operation_recorder_bp.route('/api/recorder/<record_id>', methods=['GET'])
def get_record(record_id):
    """录制详情"""
    if _demo_runner:
        record = _demo_runner.get_record(record_id)
        if not record:
            return jsonify({'success': False, 'message': '录制不存在'}), 404
        return jsonify({'success': True, 'data': record})
    return jsonify({'success': False}), 500


@operation_recorder_bp.route('/api/recorder/<record_id>/replay', methods=['POST'])
def replay(record_id):
    """回放"""
    data = request.get_json() or {}
    speed = data.get('speed', 1.0)
    if _demo_runner:
        result = _demo_runner.replay(record_id, speed)
        return jsonify({'success': True, 'data': result})
    return jsonify({'success': False}), 500


@operation_recorder_bp.route('/api/recorder/<record_id>/export', methods=['GET'])
def export_record(record_id):
    """导出录制"""
    if not _demo_runner:
        return jsonify({'success': False}), 500
    record = _demo_runner.get_record(record_id)
    if not record:
        return jsonify({'success': False, 'message': '录制不存在'}), 404

    if _export_engine:
        result = _export_engine.export_to_json(record, f'record_{record_id}.json')
        file_path = result.get('file_path')
        if file_path:
            return send_file(file_path, as_attachment=True)

    return jsonify({'success': False, 'message': '导出失败'}), 500


@operation_recorder_bp.route('/api/recorder/import', methods=['POST'])
def import_record():
    """导入录制"""
    if 'file' not in request.files:
        return jsonify({'success': False, 'message': '无文件'}), 400
    return jsonify({'success': True, 'message': '导入成功'})


@operation_recorder_bp.route('/api/recorder/<record_id>', methods=['DELETE'])
def delete_record(record_id):
    """删除录制"""
    if _demo_runner:
        return jsonify({'success': True})
    return jsonify({'success': False}), 500

File: frontend/js/ui/operation-recorder.js
/**
 * 操作捕捉回放器
 * 受 GPL v3.0 保护
 */
class OperationRecorder {
    constructor(demoPlayer) {
        this.demoPlayer = demoPlayer;
        this.recording = false;
        this.replaying = false;
        this.records = [];
        this.currentRecord = null;
    }

    init() { this._loadFromStorage(); }

    startRecord() {
        this.recording = true;
        this.currentRecord = {
            id: `REC-${Date.now()}`,
            actions: [],
            startTime: Date.now(),
        };
        document.addEventListener('click', this._captureAction.bind(this));
        document.addEventListener('input', this._captureAction.bind(this));
        document.addEventListener('change', this._captureAction.bind(this));
        alert('开始录制');
    }

    stopRecord() {
        this.recording = false;
        document.removeEventListener('click', this._captureAction.bind(this));
        document.removeEventListener('input', this._captureAction.bind(this));
        document.removeEventListener('change', this._captureAction.bind(this));
        if (this.currentRecord && this.currentRecord.actions.length > 0) {
            this.records.push(this.currentRecord);
            this._saveToStorage();
        }
        alert('录制完成');
        return this.currentRecord ? this.currentRecord.id : null;
    }

    recordAction(action) {
        if (!this.currentRecord) return;
        this.currentRecord.actions.push({
            timestamp: Date.now(),
            action,
        });
    }

    saveRecord(recordId) { this._saveToStorage(); return { success: true, record_id: recordId }; }
    getRecords() { return this.records; }
    loadRecord(recordId) { return this.records.find(r => r.id === recordId); }

    async replay(recordId, speed = 1.0) {
        const record = this.loadRecord(recordId);
        if (!record) return;
        this.replaying = true;
        const startTime = record.startTime;
        for (const action of record.actions) {
            if (!this.replaying) break;
            const delay = (action.timestamp - startTime) / speed;
            await this._wait(delay);
            this._executeAction(action);
        }
        this.replaying = false;
    }

    pause() { this.replaying = false; }
    resume() { this.replaying = true; }
    stop() { this.replaying = false; }

    deleteRecord(recordId) {
        this.records = this.records.filter(r => r.id !== recordId);
        this._saveToStorage();
    }

    exportRecord(recordId) {
        const record = this.loadRecord(recordId);
        if (!record) return;
        const blob = new Blob([JSON.stringify(record)], { type: 'application/json' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = `${recordId}.json`;
        a.click();
    }

    importRecord(file) {
        const reader = new FileReader();
        reader.onload = (e) => {
            try {
                const record = JSON.parse(e.target.result);
                this.records.push(record);
                this._saveToStorage();
            } catch (err) {
                alert('导入失败：' + err.message);
            }
        };
        reader.readAsText(file);
    }

    _captureAction(event) {
        if (!this.recording) return;
        this.recordAction({
            type: event.type,
            target: event.target.id || event.target.tagName,
            data: {
                x: event.clientX,
                y: event.clientY,
                value: event.target.value,
            },
        });
    }

    _executeAction(action) {
        const a = action.action;
        switch (a.type) {
            case 'click': {
                const el = document.getElementById(a.target);
                if (el) el.click();
                break;
            }
            case 'input': {
                const el = document.getElementById(a.target);
                if (el) {
                    el.value = a.data.value;
                    el.dispatchEvent(new Event('input'));
                }
                break;
            }
        }
    }

    _wait(ms) { return new Promise(resolve => setTimeout(resolve, ms)); }
    _saveToStorage() { StorageUtils.set('operation_records', this.records); }
    _loadFromStorage() { this.records = StorageUtils.get('operation_records', []); }
}

window.OperationRecorder = OperationRecorder;
document.addEventListener('DOMContentLoaded', () => {
    if (window.demoPlayer) {
        window.operationRecorder = new OperationRecorder(window.demoPlayer);
    }
});